### PR TITLE
feat: Implements settlement report time series factory

### PR DESCRIPTION
**Pull request by jpm-halfspace**
# Description

Implements functionality to generate parameterised time series for settlement reporting.

Closes Issue: https://app.zenhub.com/workspaces/taskforce-settlement-reports-66d955bb0ac43f0036e87e5c/issues/gh/energinet-datahub/team-raccoons/802

## Pull-request quality

- [x] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [x] Tests are written and executed locally
- [ ] Subsystem tests have been tested (by manually deploying to `dev_002`)

### CODE REVIEW

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 1]
```diff
@@ -11,15 +11,239 @@
...
+    if calculation_id is not None:
+        df = df.filter(F.col("calculation_id") == calculation_id)
+    else:
+        df = df.filter(F.col("calculation_id").isNull())
+
+    return (
+        df.withColumn(
+            "start_of_day",
+            F.date_trunc(
+                "DAY", F.from_utc_timestamp("observation_time", "Europe/Copenhagen")
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:102]: pass time_zone as parameter? It is already part of SettlementReportArgs

File: source/databricks/calculation_engine/package/settlement_report_job/constants.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+METERING_POINT_TIME_SERIES_VIEW = (
+    "ctl_shres_d_we_003.wholesale_settlement_reports.metering_point_time_series_v1"
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/constants.py:2]: Catalog name must be provided by infrastructure.
See [this](https://github.com/Energinet-DataHub/dh3-infrastructure/blob/main/wholesale/terraform/main/dbwj-calculator.tf), and how it's used in the calculator job (I can show you if you want).

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 3]
```diff
@@ -11,15 +11,239 @@
...
+            [
+                T.StructField("id", T.StringType(), False),
+                T.StructField("observation_time", T.TimestampType(), False),
+                T.StructField("quantity", T.DoubleType(), True),
+            ]
+        )
+    )
+)
+def _pad_quantities(quantities: list[Row], desired_obs: int) -> Column:
+    """Pad the quantities with None values to the desired number of observations.
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:48]: Kudos for writing good code doc when appropriate 👍

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 4]
```diff
@@ -11,15 +11,239 @@
...
+
+    To account for daylight saving time, we have to ensure that an extra hour is added to the
+    time series. This is done by adding an extra observation at the end of the time series
+    with a None value.
```
[the remark this answers is not in the document]
Comment (jpm-halfspace) [reply source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:52]: Tried to make the intent a bit more clear

File: source/databricks/calculation_engine/package/settlement_report_job/constants.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    "total_consumption": "D12",
+    "electrical_heating": "D14",
+    "net_consumption": "D15",
+    "effect_settlement": "D19",
+    "consumption": "E17",
+    "production": "E18",
+    "exchange": "E20",
+}
+
+WORST_CASE_GRID_AREA_CODES = {
```
[the remark this answers is not in the document]
Comment (jpm-halfspace) [reply source/databricks/calculation_engine/package/settlement_report_job/constants.py:26]: It does, I just didn't know where to put it right now

[the remark this answers is not in the document]
Comment (JesperGomes) [reply source/databricks/calculation_engine/package/settlement_report_job/constants.py:32]: Bjarke, it is because this PR focused on generating csvs for only the worst case scenario. I have a PR coming up which enables setting these values as job parameters.

[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/settlement_report_job/constants.py:32]: Ok, no problem 😄

File: source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+class TimeSeriesPointColumnNames:
+    metering_point_id = "METERINGPOINTID"
+    metering_point_type = "TYPEOFMP"
+    start_of_day = "STARTDATETIME"
+    energy_prefix = "ENERGYQUANTITY"
+
+
+class ComputedColumns:
+    uid = "uid"
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py:20]: What is this? Can we add a docstring?

File: source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    observation_time = "observation_time"
+    resolution = "resolution"
+    calculation_type = "calculation_type"
+    calculation_id = "calculation_id"
+    metering_point_id = "metering_point_id"
+    metering_point_type = "metering_point_type"
+    quantity = "quantity"
+
+
+class TimeSeriesPointColumnNames:
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py:12]: Column names in view or CSV?
```suggestion
class TimeSeriesPointCsvColumnNames:
```

File: source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+class TableColumnNames:
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py:1]: We should be reading from the standard data product. So it's more likely view column names. But I'm not sure that that will lead us to the best name here. Perhaps like this...?
```suggestion
class DataProductColumnNames:
```

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 9]
```diff
@@ -11,15 +11,353 @@
...
+            src = csv_files[0].path
+            dst = f"{path}/{resolution_name}_{grid_area_name}_{start_date_name}_{end_date_name}_{split_name}.csv"  # noqa
+            files_to_zip.append(dst)
+            dbutils.fs.mv(src, dst)
+            dbutils.fs.rm(grid_file.path, recurse=True)
+
+    return files_to_zip
+
+
+def create_time_series(
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:330]: We usually locate the public symbols at the top of modules. But I'm not sure what is best practice.

File: source/databricks/calculation_engine/package/settlement_report_job/utils.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    Raises:
+        Exception: If there are no files to zip.
+        Exception: If the save path does not end with .zip.
+    """
+    if len(files_to_zip) == 0:
+        raise Exception("No files to zip")
+    elif not save_path.endswith(".zip"):
+        raise Exception("Save path must end with .zip")
+
+    tmp_path = f"/tmp/{generate_uuid()}.zip"
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/utils.py:73]: How about replacing `generate_uuid()` with a `report_id` argument?

File: source/databricks/calculation_engine/tests/settlement_report_job/unit_tests/test_settlement_report_utils.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+from pyspark.sql import SparkSession, functions as F
+
+from package.settlement_report_job.utils import (
+    rowbind_list_of_dataframes,
+    map_from_dict,
+)
+
+
+def test_rowbind_list_of_dataframes(spark: SparkSession):
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/settlement_report_job/unit_tests/test_settlement_report_utils.py:9]: [Here](https://energinet.atlassian.net/wiki/spaces/D3/pages/648052739/The+Basics#Writing-Tests) is the documentation on our test naming convention.

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 12]
```diff
@@ -11,15 +11,353 @@
...
+        )
+    )
+)
+def _pad_quantities(quantities: list[Row], desired_obs: int) -> list[dict]:
+    """Pad the quantities with None values to the desired number of observations.
+
+    To account for daylight saving time, we have to ensure that an extra hour is
+    added to the time series. Thus, for hourly data you should set desired_obs to 25,
+    and for quarterly data you should set desired_obs to 25 * 4. The extra hour is added
+    to the end of the time series.
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:62]: Where can I see this in the code? And should it be unit-tested?

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py (hunk 9 above)
Comment (jpm-halfspace) [reply source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:328]: A matter of preference I guess. I'll move it to the top

File: source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py (hunk 6 above)
Comment (jpm-halfspace) [reply source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py:23]: I changed the name to `EphemeralColumns` and added a docstring to say that these are columns used in calculations, but ones that are not present in neither input or output

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/settlement_report_job/table_column_names.py:23]: Btw, what is "uid"? Is it just an ID? If so we'll probably just use "id" as that's what we're used to.

File: source/databricks/calculation_engine/package/settlement_report_job/constants.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    "net_consumption": "D15",
+    "effect_settlement": "D19",
+    "consumption": "E17",
+    "production": "E18",
+    "exchange": "E20",
+}
+
+RESOLUTION_NAMES = {"PT1H": "TSSD60", "PT15M": "TSSD15"}
+
+WORST_CASE_GRID_AREA_CODES = {
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/settlement_report_job/constants.py:35]: we now have the dictionary as input parameter, so we don't need to hard code it anymore

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 14]
```diff
@@ -11,15 +11,370 @@
...
+            f"{TimeSeriesPointCsvColumnNames.energy_prefix}{i+1}"
+        )
+
+    return quantity_dicts
+
+
+def _filter_metering_points(
+    metering_point_time_series: DataFrame,
+    grid_area: str,
+    calculation_type: CalculationType,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:143]: we don't need to filter by calculation_type when we filter on calculation_id.

File: source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py [hunk 15]
```diff
@@ -11,15 +11,370 @@
...
+) -> DataFrame:
+    df = metering_point_time_series.filter(
+        (F.col(DataProductColumnNames.grid_area_code) == grid_area)
+        & (F.col(DataProductColumnNames.resolution) == resolution)
+        & (F.col(DataProductColumnNames.calculation_type) == calculation_type.value)
+        & (F.col(DataProductColumnNames.observation_time) >= start_period)
+        & (F.col(DataProductColumnNames.observation_time) < end_period)
+    )
+
+    if calculation_id is not None:
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/settlement_report_job/time_series_factory.py:158]: why is this needed? Calculation id should never be None - except when we come to balance fixing, but then several other things change also.

File: source/databricks/calculation_engine/package/settlement_report_job/constants.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+
+METERING_POINT_TYPE_DICT = {
+    "ve_production": "D01",
+    "net_production": "D05",
+    "supply_to_grid": "D06",
+    "consumption_from_grid": "D07",
+    "wholesale_services_information": "D08",
+    "own_production": "D09",
+    "net_from_grid": "D10",
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/settlement_report_job/constants.py:23]: what about "net_to_grid"= "D11"?

**Review by JesperGomes — APPROVED** [review]
